Add diagnostics and formatter for Puppet

jovandeginste opened the pull request:
Based on the Rubocop formatter. This uses puppet-lint.

Signed-off-by: Jo Vandeginste <Jo.Vandeginste@kuleuven.be>

jovandeginste commented:
I have no experience whatsoever with Lua; please guide me where this PR could improve!

jovandeginste commented:
(The check seems to fail on a file I didn't touch?)

jose-elias-alvarez commented:
This seems to be based off a *really* old commit. Everything seems fine, but could you please rebase it from the latest master so that you can verify that it still works as expected (and so that I can test it out)?

jovandeginste commented:
Nov 21st is "_really_ old"? :rofl: Anyway, I rebased, and updated some things. Also added epuppet (`epp` files). Still, happy to learn where this could improve! Especially the loop is awkward I think. The "output" from `puppet-lint` is something like:

```json
[
  [
    {
      "message": "double quoted string containing no variables",
      "line": 11,
      "column": 6,
      "token": "#<PuppetLint::Lexer::Token:0x0000563212090a20>",
      "kind": "warning",
      "check": "double_quoted_strings",
      "fullpath": "/full/path/to/file.pp",
      "path": "path/to/file.pp",
      "filename": "file.pp",
      "KIND": "WARNING"
    },
    { (other issue },
  ],
  [ (other file) ],
]
```

Note the nested array; every file has a new entry, but we only check one file - should I still account for multiple files?

jovandeginste commented:
I updated the code - would you check if this is good? I didn't mind the multiple files yet...

jovandeginste commented:
Appologies - didn't update the actual branch...

jovandeginste commented:
I tried a few iterations on the multiple files approach, but couldn't get it to work properly; when I open multiple files, it seems to run the linter again when switching to the other buffers (although it should have received all linting issues with the first run).

jovandeginste commented:
I also see this message: `Client 3 quit with exit code 127 and signal 0`
I assume I'm still missing some detail somewhere; is this maybe coming from null-ls (and can you educate me?), or should I be looking at different parts of LunarVim?

jovandeginste commented:
If this is good, merge ahead!

jose-elias-alvarez commented:
Thanks!
